x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/cache/CacheServiceTests.java

```diff
@@ -221,44 +226,150 @@ public void testPut() throws Exception {
         }
     }
 
-    public void testRunIfShardMarkedAsEvictedInCache() throws Exception {
-        final SnapshotId snapshotId = new SnapshotId(randomAlphaOfLength(5).toLowerCase(Locale.ROOT), UUIDs.randomBase64UUID(random()));
-        final IndexId indexId = new IndexId(randomAlphaOfLength(5).toLowerCase(Locale.ROOT), UUIDs.randomBase64UUID(random()));
-        final ShardId shardId = new ShardId(randomAlphaOfLength(5).toLowerCase(Locale.ROOT), UUIDs.randomBase64UUID(random()), 0);
-        final Path cacheDir = Files.createDirectories(resolveSnapshotCache(randomShardPath(shardId)).resolve(snapshotId.getUUID()));
-
+    public void testMarkShardAsEvictedInCache() throws Exception {
         final CacheService cacheService = defaultCacheService();
-        cacheService.setCacheSyncInterval(TimeValue.ZERO);
         cacheService.start();
 
-        cacheService.runIfShardMarkedAsEvictedInCache(
-            snapshotId,
-            indexId,
-            shardId,
-            () -> { assert false : "should not be called: shard is not marked as evicted yet"; }
-        );
+        final List<CacheFile> randomCacheFiles = randomCacheFiles(cacheService);
+        assertThat(cacheService.pendingShardsEvictions(), aMapWithSize(0));
+
+        final ShardEviction shard = randomShardEvictionFrom(randomCacheFiles);
+        final List<CacheFile> cacheFilesAssociatedWithShard = filterByShard(shard, randomCacheFiles);
+        cacheFilesAssociatedWithShard.forEach(cacheFile -> assertTrue(Files.exists(cacheFile.getFile())));
+
+        final BlockingEvictionListener blockingListener = new BlockingEvictionListener();
+        final CacheFile randomCacheFile = randomFrom(cacheFilesAssociatedWithShard);
+        assertTrue(Files.exists(randomCacheFile.getFile()));
+        randomCacheFile.acquire(blockingListener);
+
+        for (int i = 0; i < between(1, 3); i++) {
+            cacheService.markShardAsEvictedInCache(shard.getSnapshotUUID(), shard.getSnapshotIndexName(), shard.getShardId());
```

Comment:
Should we maybe randomly also delete one (or all) of the files in the shard before calling this?

Reply:
That's a good idea, I pushed aae2f78a7eda3dc1767897c51c877238afc2b075